File: app/src/main/java/com/itba/runningMate/pastruns/runs/RunAdapter.java
package com.itba.runningMate.pastruns.runs;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.itba.runningMate.R;
import com.itba.runningMate.domain.Run;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import static androidx.recyclerview.widget.RecyclerView.NO_ID;

public class RunAdapter extends RecyclerView.Adapter<RunViewHolder> {

    private final List<Run> currentRunList;
    private WeakReference<OnRunClickListener> listener;

    public RunAdapter() {
        currentRunList = new ArrayList<>();
    }

    public void update(List<Run> runList) {
        this.currentRunList.clear();
        if (runList != null) {
            this.currentRunList.addAll(runList);
        }
        notifyDataSetChanged();
    }

    public void setClickListener(OnRunClickListener listener) {
        this.listener = new WeakReference<>(listener);
    }

    @Override
    public int getItemViewType(final int position) {
        return R.layout.run_element;
    }

    @NonNull
    @Override
    public RunViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
        return new RunViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RunViewHolder holder, int position) {
        if (listener == null) {
            return;
        }
        holder.bind(currentRunList.get(position));
        holder.setOnClickListener(listener.get());
    }

    @Override
    public long getItemId(int position) {
        if (currentRunList.isEmpty()) {
            return NO_ID;
        }

        Run pos = currentRunList.get(position);

        return pos == null ? NO_ID : pos.getUid();
    }

    @Override
    public int getItemCount() {
        return currentRunList.size();
    }

}

File: app/src/main/java/com/itba/runningMate/pastruns/runs/RunAdapter.kt
package com.itba.runningMate.pastruns.runs

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.itba.runningMate.components.run.OnRunClickListener
import com.itba.runningMate.components.run.RunElementView
import com.itba.runningMate.domain.Run
import java.lang.ref.WeakReference
import java.util.*

class RunAdapter : RecyclerView.Adapter<RunViewHolder>() {

    private val currentRunList: MutableList<Run>
    private var clickListener: WeakReference<OnRunClickListener>? = null
    private var deleteListener: WeakReference<OnRunDeleteListener>? = null

    fun update(runList: List<Run>?) {
        currentRunList.clear()
        if (runList != null) {
            currentRunList.addAll(runList)
        }
        notifyDataSetChanged()
    }

    fun setClickListener(listener: OnRunClickListener) {
        this.clickListener = WeakReference(listener)
    }

    fun setSwipeRunToDeleteListener(listener: OnRunDeleteListener) {
        this.deleteListener = WeakReference(listener)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RunViewHolder {
        val view = RunElementView(parent.context)
        view.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        return RunViewHolder(view)
    }

    override fun onBindViewHolder(holder: RunViewHolder, position: Int) {
        if (clickListener == null) {
            return
        }
        holder.setOnClickListener(clickListener!!.get()!!)
        holder.bind(currentRunList[position])
    }

    override fun getItemId(position: Int): Long {
        if (currentRunList.isEmpty()) {
            return RecyclerView.NO_ID
        }
        val pos = currentRunList[position]
        return pos.uid ?: RecyclerView.NO_ID
    }

    override fun getItemCount(): Int {
        return currentRunList.size
    }

    fun deleteItem(position: Int) {
        val runId = currentRunList[position].uid!!
        currentRunList.removeAt(position)
        notifyDataSetChanged()
        if (deleteListener?.get() != null) {
            deleteListener!!.get()!!.onSwipeRunToDelete(runId)
        }
    }

    init {
        currentRunList = ArrayList()
    }
}